src/Passes/site-packages/TasksCI/builder.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import os
+from . import settings
+from . import toolchain
+from .settings import PROJECT_ROOT
+import logging
+import subprocess
+import sys
+from typing import Union, List
+
+OptionalInt = Union[int, None]
+OptionalStr = Union[str, None]
+
+
+logger = logging.getLogger()
+
+
+def configure_cmake(build_dir: str, generator=None, cmake_flags: List[str] = []) -> None:
+    """
+    Function that creates a build directory and runs
+    cmake to configure make, ninja or another generator.
+    """
+
+    logger.info("Source: {}".format(PROJECT_ROOT))
+    logger.info("Build : {}".format(build_dir))
+
+    os.chdir(PROJECT_ROOT)
+    os.makedirs(build_dir, exist_ok=True)
+
+    cmake_cmd = [toolchain.discover_cmake()]
+
+    if generator is not None:
+        cmake_cmd += ['-G', generator]
+
+    cmake_cmd += [PROJECT_ROOT]
+    cmake_cmd += cmake_flags
+
+    exit_code = subprocess.call(cmake_cmd, cwd=build_dir)
+    if exit_code != 0:
+        logger.error('Failed to configure project')
+        sys.exit(exit_code)
+
+
+def build_project(build_dir: str, generator: OptionalStr = None, concurrency: OptionalInt = None) -> None:
+    """
+    Given a build directory, this function builds all targets using
+    a specified generator and concurrency.
+    """
+
+    if generator in ["make", None]:
+        cmd = ["make"]
+    elif generator in ["ninja"]:
+        cmd = ["ninja"]
+
+    if concurrency is None:
+        concurrency = settings.get_degree_of_concurrency()
+
+    cmd.append('-j{}'.format(concurrency))
+
+    exit_code = subprocess.call(cmd, cwd=build_dir)
+
+    if exit_code != 0:
+        logger.error('Failed to make the project')
+        sys.exit(-1)
+
+
+def run_tests(build_dir: str, concurrency: OptionalInt = None) -> None:
+    """
+    Runs the unit tests given a build directory.
+    """
+    fail = False
+
+    # Lit is disabled for the time being
+    if False:
+        # Running lit tests
+        lit_cmd = ["lit", "tests/", "-v"]
+        exit_code = subprocess.call(lit_cmd,  cwd=build_dir)
+
+        if exit_code != 0:
+            logger.error('Lit test failed')
+            fail = True
+
+    # Running CMake tests
+    cmake_cmd = [toolchain.discover_ctest()]
+
+    if concurrency is not None:
+        raise BaseException("No support for concurrent testing at the moment.")
+
+    exit_code = subprocess.call(cmake_cmd, cwd=build_dir)
+    if exit_code != 0:
+        logger.error('CTest failed project')
+
+        # Getting output of failed
+        cmake_cmd.append("--rerun-failed")
+        cmake_cmd.append("--output-on-failure")
+        subprocess.call(cmake_cmd, cwd=build_dir)
+
+        fail = True
+
+    if fail:
+        sys.exit(exit_code)
+
+
+def main(build_dir: str, generator: OptionalStr = None, test: bool = False, cmake_flags: List[str] = []) -> None:
```

Comment:
To my knowledge, this repository does not already use PSake, Cake or Invoke which means it would just be throwing another build system into the mix. Other solutions were considered such as writing this script in PowerShell as well as whether these tasks could be handled directly within CMake. At the end, a pure Python solution was chosen as it was the simplest solution that could be executed the quickest. In time, the build script should be moved to PowerScript.